src/inmanta/server/services/environmentservice.py

```diff
@@ -524,11 +530,18 @@ async def environment_clear(self, env: data.Environment) -> None:
         if is_protected_environment:
             raise Forbidden(f"Environment {env.id} is protected. See environment setting: {data.PROTECTED_ENVIRONMENT}")
 
-        await self.autostarted_agent_manager.stop_agents(env, delete_venv=True)
+        await asyncio.gather(
+            self.autostarted_agent_manager.stop_agents(env, delete_venv=True),
+            self.compiler_service.cancel_compile(env.id),
```

Comment:
Isn't there a race condition here in the case where there are multiple compiles queued up ? We cancel the currently running one but couldn't the next one start right after ?
I must admit I'm not too clear on the semantics between clear and delete and
- why we need the env to be halted when deleting but not when clearing ?
- why we don't need the `environment_state_operation_lock` when clearing ?

Reply:
> Isn't there a race condition here in the case where there are multiple compiles queued up ?

No. Because the compile that finishes starts the next one. But there is one if a new compile is requested right after the previous one is cancelled. That's something I need to fix.

> why we need the env to be halted when deleting but not when clearing ?

I agree it would be better to halt the environment right before we clear is, but that would be a breaking change on the API. Because currently a clear of an environment doesn't halt it. But maybe I can work around that.

> why we don't need the environment_state_operation_lock when clearing ?

It protects against race conditions that would result in an inconsistent state on the agent state (paused or not) and the environment state (halted or not). The clear endpoint doesn't change the environment state, so there is no need for the lock.

Reply:
> Because the compile that finishes starts the next one.

But isn't this an issue ? i.e. deleting the env dir and clearing the env while there is an ongoing compile ?

Reply:
I mean that the normal working of the compiler service is to start a new compile (if any is present in the queue) when a compile finishes. The `cancel_compile()` explicitly deviates from that to make sure that no compile is executing. But it's not waterproof, as mentioned in the your first comment. I will fix this.

Reply:
I completely misunderstood what self._recompiles was. For some reason I thought it was a queue of pending compiles, sorry for the noise. I see now how popping it in `cancel_compile()` prevents further compiles from being queued up.